### PR TITLE
Intersections of rays

### PR DESCRIPTION
**Pull request by dorn-gerhard**
This PR adds all possible intersections of 2D and 3D rays using the new IntersectionTypes:
CrossingRays
MidTouchingRays
CornerTouchingRays
OverlappingAlignedRays
OverlappingCollidingRays

### CODE REVIEW

File: src/intersections.jl [hunk 1]
```diff
@@ -44,6 +44,9 @@ The different types of intersection that may occur between geometries.
 Type `IntersectionType` in a Julia session to see the full list.
 """
 @enum IntersectionType begin
+  # no intersection
+  NoIntersection
```
Comment (dorn-gerhard) [inline src/intersections.jl:48]: `NoIntersection` is now linked to zero

File: src/intersections/rays.jl [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+5. overlap with colliding vectors (OverlappingCollidingRays -> Segment)
+6. do not overlap nor intersect (NoIntersection)
+=#
+function intersection(f, r1::Ray{3,T}, r2::Ray{3,T}) where {T}
+  a, b = r1(0), r1(1)
+  c, d = r2(0), r2(1)
+
+  if measure(Tetrahedron(a, b, c, d)) > 0 # not in same plane
+    return @IT NoIntersection nothing f #CASE 6
+  elseif isapprox(norm((b - a) × (c - d)), zero(T), atol=atol(T)^2) # parallel
```
Comment (dorn-gerhard) [inline src/intersections/rays.jl:21]: 2D and 3D versions only differs in lines 19-21.
Not sure how to resolve this redundancy - open for suggestions

File: test/intersections.jl [hunk 3]
```diff
@@ -150,6 +150,34 @@
     @inferred someornone(s1, s2)
   end
 
+  @testset "Rays" begin
```
Comment (dorn-gerhard) [inline test/intersections.jl:153]: tests for 2D
image

Please give feedback for 2D tests, then I will add tests for 3D

What else has to be tested?

**Review by dorn-gerhard — COMMENTED** [review]
I added tests for 2D ray intersection, 3D tests will follow later
hasintersect not implemented

Note sure how to correctly implement type stability tests:
# type stability tests
   r1 = Ray(..., ...)
   r2 = Ray(..., ...)
    @inferred someornone(s1, s2)

File: src/Meshes.jl [hunk 4]
```diff
@@ -302,23 +302,38 @@ export
 
   # intersection types
   IntersectionType,
+  
+  # no intersection
+  NoIntersection,
+  # segment-segment intersection
```
Comment (juliohm) [inline src/Meshes.jl:308]: This change should not be here

File: src/Meshes.jl [hunk 5]
```diff
@@ -302,23 +302,38 @@ export
...
   TouchingRayBox,
+  # segment-triangle intersection
   IntersectingSegmentTriangle,
+  # ray-triangle intersection
   IntersectingRayTriangle,
+  # segment-plane intersection
   CrossingSegmentPlane,
   TouchingSegmentPlane,
   OverlappingSegmentPlane,
-  NoIntersection,
```
Comment (juliohm) [inline src/Meshes.jl:321]: All these changes seem unrelated to the PR

File: src/intersections.jl (hunk 1 above)
Comment (juliohm) [reply src/intersections.jl:48]: I was already linked to zero in the latest master version. Please start your PR based on the latest master branch.

File: test/intersections.jl (hunk 3 above)
Comment (juliohm) [reply test/intersections.jl:153]: This is a great set of rays to test 👍🏽

File: test/intersections.jl [hunk 6]
```diff
@@ -150,6 +150,38 @@
...
+    r₄ = Ray(P2(0,2), V2(1,-3))
+    r₅ = Ray(P2(4,1.5), V2(4,2))
+    r₆ = Ray(P2(2,0.5), V2(-0.5, -0.25))
+    r₇ = Ray(P2(4,0), V2(0,1))
+    @test intersection(r₁, r₂) |> type == CrossingRays #CASE 1
+    @test r₁ ∩ r₂ == P2(1.25,0.125)
+    @test r₁ ∩ r₇ == P2(4, 1.5)
+    @test intersection(r₁, r₃) |> type == MidTouchingRays #CASE 2
+    @test r₁ ∩ r₃ == r₁.p #origin of first ray
+    @test r₅ ∩ r₇ == r₅.p
```
Comment (juliohm) [inline test/intersections.jl]: We never access fields of structs directly. We can only use the exported API, in this case origin(ray) gives the origin point.

**Review by juliohm — CHANGES_REQUESTED** [review]
@dorn-gerhard I suggest starting the PR from the latest master branch, it seems to contain changes that are already present therein.

File: src/Meshes.jl (hunk 4 above)
Comment (dorn-gerhard) [reply src/Meshes.jl:308]: I tried to undo all the changes to start from the original master file. 
Not sure if it worked.

File: src/Meshes.jl (hunk 5 above)
Comment (dorn-gerhard) [reply src/Meshes.jl:321]: see above

File: src/intersections.jl (hunk 1 above)
Comment (dorn-gerhard) [reply src/intersections.jl:48]: I tried to undo all my local changes and start from the original master file. Hope it worked.

**Review by dorn-gerhard — COMMENTED** [review]
I tried to start from the original intersections.jl file from the master branch. 

Not sure if I did it correctly

**Comment by juliohm** [comment]
@dorn-gerhard try to start from scratch. Run `git pull` on the master branch and start a fresh branch from it. This PR is polluted with changes from older versions. It is easier to start clean. You just need to copy paste the latest versions of the files you have in this new branch.
